diff -r 7af530381fec runtime/indent/ocaml.vim
--- a/runtime/indent/ocaml.vim	Sun Aug 17 17:24:07 2014 +0200
+++ b/runtime/indent/ocaml.vim	Thu Sep 11 08:28:31 2014 +0100
@@ -8,9 +8,6 @@
 "               2005 Jun 25 - Fixed multiple bugs due to 'else\nreturn ind' working
 "               2005 May 09 - Added an option to not indent OCaml-indents specially (MM)
 "               2013 June   - commented textwidth (Marc Weber)
-"
-" Marc Weber's comment: This file may contain a lot of (very custom) stuff
-" which eventually should be moved somewhere else ..
 
 " Only load this indent file when no other was loaded.
 if exists("b:did_indent")
@@ -18,15 +15,11 @@
 endif
 let b:did_indent = 1
 
-setlocal expandtab
 setlocal indentexpr=GetOCamlIndent()
 setlocal indentkeys+=0=and,0=class,0=constraint,0=done,0=else,0=end,0=exception,0=external,0=if,0=in,0=include,0=inherit,0=initializer,0=let,0=method,0=open,0=then,0=type,0=val,0=with,0;;,0>\],0\|\],0>},0\|,0},0\],0)
 setlocal nolisp
 setlocal nosmartindent
 
-" At least Marc Weber and Markus Mottl do not like this:
-" setlocal textwidth=80
-
 " Comment formatting
 if !exists("no_ocaml_comments")
  if (has("comments"))
